### PR TITLE
feat: support for building workspaces and diff-engine inside docker

### PR DESCRIPTION
**Pull request by notnmeyer**
## What
Adds support for building npm packages and diff-engine (except macos) inside docker (and maintains support for building them directly on your workstation. I'd been working ton these changes in the context of some CI improvements in https://github.com/opticdev/optic/compare/build-in-container. I split these bits out because I think there's standalone benefit to this flexibility. 

 Build directly on your workstation,
```
➜ task workspaces:build diff-engine:add-targets diff-engine:build-linux diff-engine:build-windows
```

Build nside our devel container,
```
➜ docker-compose run devel \
  task workspaces:build diff-engine:add-targets diff-engine:build-linux diff-engine:build-windows
```

The devel container mounts the repo as a volume at /app inside. So filesystem changes within that path are persisted (think node_modules). Word of caution, running `yarn install`/`workspaces:setup` will cause  node_modules will be reinstalled each time  you switch running it in/out of docker. This is fixed in Yarn v2, so it's more something to be aware of than a real problem for now.

## Why
Flexibility in how and where this builds.

- A development image can make it easier to get up and running with our stuff without the need for brittle local configurations.
- Docker-based CI workflows give us additional options for caching and sharing data between workflows or runs.

## Validation
* [x] The tasks run. Both example commands above run for me and exit 0.
* [x] CI workflows that use the build tasks still work
    * [x] regression suite, https://github.com/opticdev/optic/runs/2116662288?check_suite_focus=true
    * [x] side-channel, https://github.com/opticdev/optic/actions/runs/655860156/workflow

### CODE REVIEW

File: docker/private-npm-registry/docker-compose.yml [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
-volumes:
-  verdaccio:
-    driver: local
+      - ./conf:/verdaccio/conf
+    networks:
+      - optic
+
+networks:
+  optic:
+    name: optic
```
Comment (devdoshi) [inline docker/private-npm-registry/docker-compose.yml:16]: my linux box complained about this field, removing it worked for me

**Comment by devdoshi** [comment]
```
$ AWS_SDK_LOAD_CONFIG=true AWS_PROFILE=staging docker-compose run devel   task workspaces:build diff-engine:add-targets diff-engine:build-linux diff-engine:build-windows
task: yarn install
yarn install v1.22.10
[1/4] Resolving packages...
[2/4] Fetching packages...
warning Pattern ["prop-types@latest"] is trying to unpack in the same destination "/home/appuser/.cache/yarn/v6/npm-prop-types-15.7.2-52c41e75b8c87e72b9d9360e0206b99dcbffa6c5-integrity/node_modules/prop-types" as pattern ["prop-types@^15.7.2","prop-types@^15.7.2","prop-types@^15.6.0","prop-types@^15.7.2","prop-types@^15.6.2","prop-types@^15.7.2","prop-types@^15.6.2","prop-types@^15.5.4","prop-types@^15.7.2","prop-types@^15.5.7","prop-types@^15.6.2","prop-types@^15.7.2","prop-types@15.7.2","prop-types@^15.7.2","prop-types@^15.7.2","prop-types@^15.7.2","prop-types@^15.6.2","prop-types@^15.6.2","prop-types@^15.6.2","prop-types@^15.7.2","prop-types@^15.7.2","prop-types@^15.7.2","prop-types@^15.6.2","prop-types@^15.5.8","prop-types@^15.6.2","prop-types@^15.5.8","prop-types@^15.6.2","prop-types@^15.6.2","prop-types@^15.7.2","prop-types@^15.7.2","prop-types@^15.6.1","prop-types@^15.6.2","prop-types@^15.6.1","prop-types@^15.6.2","prop-types@^15.6.2","prop-types@^15.7.2","prop-types@^15.6.0","prop-types@^15.6.1","prop-types@^15.7.2","prop-types@^15.6.1","prop-types@^15.6.0","prop-types@^15.6.1"]. This could result in non-deterministic behavior, skipping.
warning mini-css-extract-plugin@0.5.0: Invalid bin field for "mini-css-extract-plugin".
warning url-loader@1.1.2: Invalid bin field for "url-loader".
info fsevents@2.1.3: The platform "linux" is incompatible with this module.
info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.13: The platform "linux" is incompatible with this module.
info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
error Could not write file "/app/yarn-error.log": "EACCES: permission denied, open '/app/yarn-error.log'"
error An unexpected error occurred: "EACCES: permission denied, unlink '/app/node_modules/.yarn-integrity'".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
task: Failed to run task "workspaces:setup": exit status 1
``` is what i see on my linux box

**Comment by devdoshi** [comment]
tried again after hosing my node_modules and workspaces/*/node_modules but it failed again, seems like it doesn't like something about the permissions

File: docker/private-npm-registry/docker-compose.yml (hunk 1 above)
Comment (notnmeyer) [reply docker/private-npm-registry/docker-compose.yml:16]: that the network didnt exist? `task utility:create-docker-network`

**Comment by notnmeyer** [comment]
> error Could not write file "/app/yarn-error.log":

did you try trashing this too?

on linux there can be permissions issues on the volume mount since the files are owned by the same UID inside as outside. launch a shell, and `ls -l` the /app dir to see what i mean. you'll likely see the files owned by a numeric user.

i have an idea for working around it. will holler on slack to see if it works for you.

**Review by devdoshi — APPROVED** [review]

File: docker/private-npm-registry/docker-compose.yml (hunk 1 above)
Comment (notnmeyer) [reply docker/private-npm-registry/docker-compose.yml:16]: worked through this on slack, it was an out of date docker-compose in the ubuntu repos. installing from the docker repo resolved the issue.

**Comment by notnmeyer** [comment]
dev's permissions issues were indeed ownership related with the docker volume. we can resolve them by not dropping to `appuser` in the devel image, which im fine with as the devel image would never be running in a staging/prod env. ill follow up with a pr to modify that since this is already approved

**Pull request merged**